gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/TarArchiveInputStreamDataWriter.java

```diff
@@ -0,0 +1,120 @@
+import org.apache.hadoop.fs.Path;
+
+
+/**
+ * An {@link InputStreamDataWriter} to write archived {@link InputStream}s. The {@link #write(FileAwareInputStream)}
+ * method receives a {@link GZIPInputStream} and converts it to a {@link TarArchiveInputStream}. Each
+ * {@link TarArchiveEntry} is then written to the {@link FileSystem}.
+ */
+@Slf4j
+public class TarArchiveInputStreamDataWriter extends InputStreamDataWriter {
+
+  public TarArchiveInputStreamDataWriter(State state, int numBranches, int branchId) throws IOException {
+    super(state, numBranches, branchId);
+  }
+
+  /**
+   * Untars the passed in {@link FileAwareInputStream} to the task's staging directory. Uses the name of the root
+   * {@link TarArchiveEntry} in the stream as the directory name for the untarred file. The method also commits the data
+   * by moving the file from staging to output directory.
+   *
+   * @param fileAwareInputStream the inputStream to be written. The {@link CopyableFile} instance enclosed is used to
+   *          derive output file permissions
+   * @see #commit()
+   *
+   * @see gobblin.data.management.copy.writer.InputStreamDataWriter#write(gobblin.data.management.copy.FileAwareInputStream)
+   */
+  @Override
+  public void write(FileAwareInputStream fileAwareInputStream) throws IOException {
+    closer.register(fileAwareInputStream.getInputStream());
+
+    filesWritten++;
+
+    TarArchiveInputStream tarIn = new TarArchiveInputStream(fileAwareInputStream.getInputStream());
+    TarArchiveEntry tarEntry;
+
+    boolean firstEntrySeen = false;
+
+    try {
+      while ((tarEntry = tarIn.getNextTarEntry()) != null) {
+
+        Path tarEntryPath =
+            PathUtils.withoutLeadingSeparator(new Path(fileAwareInputStream.getFile().getDestination().getParent(),
+                tarEntry.getName()));
+
+        Path tarEntryStagingPath = new Path(this.stagingDir, tarEntryPath);
+
+        /*
+         * If this is the first entry in the tar, reset the destination file name to the name of the first entry in the
+         * tar. It is required to change the name of the destination file because the untarred file/dir name is known to
+         * the source
+         */
+        if (!firstEntrySeen) {
+          String newFileName = tarEntryStagingPath.getName();
+          fileAwareInputStream.getFile().setDestination(
+              new Path(fileAwareInputStream.getFile().getDestination().getParent(), newFileName));
+          fileAwareInputStream.getFile().setRelativeDestination(
+              new Path(fileAwareInputStream.getFile().getRelativeDestination().getParent(), newFileName));
+          firstEntrySeen = true;
+        }
+
+        log.info("Unarchiving " + tarEntryStagingPath);
+
+        if (tarEntry.isDirectory() && !fs.exists(tarEntryStagingPath)) {
+          fs.mkdirs(tarEntryStagingPath);
+        } else if (!tarEntry.isDirectory()){
+
+          FSDataOutputStream out = fs.create(tarEntryStagingPath, true);
```

Comment:
You should use https://hadoop.apache.org/docs/current/api/org/apache/hadoop/fs/FSDataInputStream.html#read%28org.apache.hadoop.io.ByteBufferPool,%20int%29 and https://docs.oracle.com/javase/7/docs/api/java/nio/channels/Channels.html#newChannel(java.io.OutputStream) to avoid unnecessary copy of data to the heap.